runtime: Add more steps to the lifecycle docs

wking opened the pull request:
And explain how the handling differs when an application joins an
existing container vs. creating a new container.  This is spun off
from the [mailing list thread][1] now that that's settled down.

This write-up is just about Linux containers, since I don't understand
the other systems well enough to do them justice.  I'm happy to roll
in notes for other platforms if folks know what to write ;).

I've pulled the hook-environment change I made in [this post][2] out
into a separate commit, and I'm happy to push that as a separate PR if
it makes review easier.

[1]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/frUXLljXy8Y
[2]: https://groups.google.com/a/opencontainers.org/d/msg/dev/frUXLljXy8Y/_sg-BidnAwAJ

mrunalp commented on runtime.md:
```diff
@@ -29,24 +29,59 @@ The root directory to the bundle is provided in the state so that consumers can
+A typical lifecyle progresses like this:
+
+1. There is no container or running application
+2. A user tells the runtime to start a container+application
```
start a container for an application.. ?

crosbymichael replied:
`A user tells the runtime to start a container`  Don't use the word application, its just a container

mrunalp replied:
Yeah, I think it makes sense to leave out application altogether.

wking replied:
On Wed, Sep 30, 2015 at 11:52:37AM -0700, Mrunal Patel wrote:
> > +2. A user tells the runtime to start a container+application
> 
> start a container for an application.. ?

Without explicit create and start steps, it's “to create a container
and launch an application inside that container”, but that seemed to
long.  Should I just replace “container+application” with “container
and application”?

wking commented:
I've rebased on the new master and added 80c1729 since #202 landed.  I'm not sure what the intended behavior isif the application dies while we're running post-start hooks.  In 80c1729 I've gone with “we terminate any in-progress post-start hook and skip any that hadn't been called yet”, but “terminate” is vague (TERM?  KILL?  TERM and KILL after 5 seconds if necessary?).  @LK4D4, do you want something different than the 80c1729 phrasing?

justincormack commented:
Is this supposed to be descriptive or normative? It might be clearer to insert normative language where useful.

Destroying the "container process" (might be a better term, its usually going to be container pid 1 but I suppose you might not have a pid namespace) will often implicitly destroy the container, so I would allow for that to happen, so hooks after that may not be able to see container any more.

wking commented:
On Thu, Oct 01, 2015 at 02:23:34AM -0700, Justin Cormack wrote:
> Is this supposed to be descriptive or normative?

I'd rather have the summaries in the enumerated list be descriptive
(so you can cut a few corners there), but the sections they link to be
normative (so everyone is on the same page for what's really
happening).

> Destroying the "container process" (might be a better term, its
> usually going to be container pid 1 but I suppose you might not have
> a pid namespace) will often implicitly destroy the container, so I
> would allow for that to happen, so hooks after that may not be able
> to see container any more.

I don't see “destroy” in the language through e0ec87b.  Are you
suggesting that language for:

  11. The runtime [terminates any other processes in the container](#stop-process)

This termination is qualified by the “Joining existing containers”
section, that clarifies the “other processes in the container” means
“other processes in any cgroup you created” (at least on Linux).  I'm
happy to roll in changes that make that more clear (maybe in [2]
itself?), if you have suggestions for the text you'd like to see.

This would also be easier if we landed a glossary like that proposed
in #107, although that doesn't currently equate “container processes”
with “processes in a unified-hierarchy cgroup” like we do in this PR.

[1]: https://github.com/wking/opencontainer-specs/blob/lifecycle/runtime.md#joining-existing-containers
[2]: https://github.com/wking/opencontainer-specs/blob/lifecycle/runtime.md#stop-process

wking commented:
I did a bit more digging in runC's cleanup implementation yesterday,
and here's what it's currently doing:

1. If the container process was the first in a PID namespace, it's
   death will automatically cleanup any other processes in that
   namespace.  Otherwise, the current runC code kills the other
   processes in the cgroup [1,2].  The ‘Freeze’ function requires a
   freezer cgroup [3], but the killer carries on if it fails [4].  The
   ‘GetPids’ function looks at the device cgroup [5].  The ‘Kill’ call
   sends a SIGKILL [6,7,8,9].  Wait for the killed processes to exit
   [10].
2. Remove any created cgroups [11]
3. Remove /run/opencontainer/containers/{container-id} [12].
4. Run the post-stop hooks [13].

I'm not sure when persistent namespaces (e.g. network namespaces) are
removed.

That's a bit different from the cleanup logic currently proposed here:

a. I suggest running the post-stop hooks before cleaning up the
   container.
b. I suggest removing the state.json file first (with the just-pushed
   e0ec87b → 717a999) to be the inverse of the container-creation
   process, where state.json is written last.
c. I require a unified hierarchy, which I thought was our intention
   after 429f936 (Adding cgroups path to the Spec, 2015-09-02, #137)
   [14].  The unified hierarchy is still experimental [15], but it
   makes handling containers easier because we don't have to create a
   PID namespace or device cgroup before we can easily define “the
   process belonging to the container”.  With the unified hierarchy
   (‘mount -t cgroup -o __DEVEL__sane_behavior cgroup $MOUNT_POINT’),
   you can add processes to a cgroup that doesn't use any controllers,
   which is what you'd want if someone left linux.devices and
   linux.resources out of their runtime.json.  I don't know of a good
   way to do this with the split hierarchy, although the devices
   controller seems like a reasonable choice for a canonical
   hierarchy, since you can join one without configuring it and keep
   the default matching the parent [16].

I think this PR's current approach to (a) and (b) makes sense, and it
would be easy to update runC to match.  I'm less clear on (c), since I
like the unified-hierarchy approach, but am cautious about relying on
an experimental kernel interface.  Maybe we should relax our
unified-hierarchy phrasing and explicitly mention the device
controller as one that must be created and whose cgroup.procs defines
the set of container processes.

[1]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/container_linux.go#L241-L245
[2]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/init_linux.go#L304-L312
[3]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/cgroups/fs/apply_raw.go#L179-L203
[4]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/init_linux.go#L310
[5]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/cgroups/fs/apply_raw.go#L205-L217
[6]: https://golang.org/pkg/os/#Process.Kill
[7]: https://golang.org/src/os/doc.go?s=1097:1127#L25
[8]: https://golang.org/src/os/exec_posix.go#L52
[9]: https://golang.org/src/os/exec_posix.go#L18
[10]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/init_linux.go#L328-L332
[11]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/container_linux.go#L246
[12]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/container_linux.go#L247-L249
[13]: https://github.com/opencontainers/runc/blob/cc84f2cc9b4e66ab85ae83546d50a74ae87fe034/libcontainer/container_linux.go#L251-L262
[14]: https://github.com/opencontainers/specs/pull/137/files/#diff-b753d6f046da0a2956b5218efd779b7fR139
[15]: https://www.kernel.org/doc/Documentation/cgroups/unified-hierarchy.txt
[16]: https://www.kernel.org/doc/Documentation/cgroups/devices.txt

crosbymichael commented on runtime-config.md:
```diff
@@ -12,27 +12,28 @@ Presently there are `Prestart`, `Poststart` and `Poststop`.
 Hooks allow one to run code before/after various lifecycle events of the container.
 Hooks MUST be called in the listed order.
 The state of the container is passed to the hooks over stdin, so the hooks could get the information they need to do their work.
+All hooks execute in the host environment (e.g. the same namespace, cgroups, etc. that apply to the host process).
```
This is incorrect.  They are forked from the runtime process so they inherit from the runtime, not all host processes.

wking replied:
On Wed, Oct 07, 2015 at 11:36:22AM -0700, Michael Crosby wrote:
> > +All hooks execute in the host environment (e.g. the same
> > namespace, cgroups, etc. that apply to the host process).
> 
> This is incorrect.  They are forked from the runtime process so they
> inherit from the runtime, not all host processes.

There are at least two runtime processes here.  There is a process
launched with (for example) ‘runC start’, which I've been referring to
as the “host process”.  That process forks and the child ends up in
the container, eventually running the config.json process binary.  I'd
been referring to the latter as the “container process”.  Do you
prefer alternative names for those two processes?

crosbymichael commented on runtime.md:
```diff
@@ -29,24 +29,60 @@ The root directory to the bundle is provided in the state so that consumers can
 
-Runs a process in a container.
-Can be invoked several times.
+After the pre-start hooks complete, the host process signals the container process to execute the runtime.
```
This sentence does not make sense.  ` the host process signals the container process to execute the runtime.`????

wking replied:
On Wed, Oct 07, 2015 at 11:40:34AM -0700, Michael Crosby wrote:
> > +After the pre-start hooks complete, the host process signals the
> > container process to execute the runtime.
>
> This sentence does not make sense.  ` the host process signals the
> container process to execute the runtime.`????

I haven't looked into the runC implementation here, but my guess was
that:

1. ‘runC start’ would launch the host process.
2. The host process would create as much of the container as it could
   without a container process to hold open things like a PID
   namespace.
3. The host process would fork, and the child of that fork would be
   the container process.
4. The container process would join the existing (partial container)
   and finish it off (e.g. by creating a PID namespace, changing
   users, dropping priveledges, etc.)
5. The host process would launch the pre-start hooks.
6. The host process would signal the container process that the
   pre-start hooks had completed successfully.
7. The container process would catch that signal and exec the
   config.json process.

The line you're quoting was supposed to sketch out (6) and (7).  Can
you clear me up on how it actually works?

wking replied:
On Wed, Oct 07, 2015 at 11:51:50AM -0700, W. Trevor King wrote:
> Wed, Oct 07, 2015 at 11:40:34AM -0700, Michael Crosby:
> > > +After the pre-start hooks complete, the host process signals
> > > the container process to execute the runtime.
> >
> > This sentence does not make sense.  ` the host process signals the
> > container process to execute the runtime.`????
> 
> I haven't looked into the runC implementation here, but my guess was
> that:
> 
> …
> 6. The host process would signal the container process that the
>    pre-start hooks had completed successfully.
> 7. The container process would catch that signal and exec the
>    config.json process.
> 
> The line you're quoting was supposed to sketch out (6) and (7).  Can
> you clear me up on how it actually works?

It looks like libcontainer actually has a fairly detailed spec for how
it sets up containers.  That talks about using a pipe instead of
singals for this trigger [1].  Signals sound easier to me (permissions
discussed in [2]), but the generic “signals” verb could cover all such
triggers (including piped messages) and not be restricted to POSIX
signals [3], since this is an internal runtime mechanism.

[1]: https://github.com/opencontainers/runc/blob/v0.0.4/libcontainer/SPEC.md#runtime-and-init-process
[2]: http://man7.org/linux/man-pages/man2/kill.2.html
[3]: http://man7.org/linux/man-pages/man7/signal.7.html

vishh commented on runtime-config.md:
```diff
@@ -12,27 +12,28 @@ Presently there are `Prestart`, `Poststart` and `Poststop`.
 ### Pre-start
 
-The pre-start hooks are called after the container process is spawned, but before the user supplied command is executed.
+The pre-start hooks are called [after the container process is spawned, but before the user supplied command is executed](runtime.md#typical-lifecycle).
```
Do we need to mention `container process` here? These hooks are invoked once the majority of the container runtime has been setup, but before the container's init process has been spawned. We need to clearly specify what sub-section of the complete container environment is available for these hooks.

wking replied:
On Mon, Oct 12, 2015 at 12:47:00PM -0700, Vish Kannan wrote:
> > -The pre-start hooks are called after the container process is spawned, but before the user supplied command is executed.
> > +The pre-start hooks are called [after the container process is spawned, but before the user supplied command is executed](runtime.md#typical-lifecycle).
> 
> Do we need to mention `container process` here? These hooks are
> invoked once the majority of the container runtime has been setup,
> but before the container's init process has been spawned. We need to
> clearly specify what sub-section of the complete container
> environment is available for these hooks.

The container is complete when the pre-start hooks run.  The container
process has been spawned (it's just running runtime-supplied code).
The only change that happens *after* the pre-start hooks is that that
container process execs the user-supplied process, swapping in the
user-selected code.  For a bit of runC-side docs on this, see the
synchronization pipe in [1].

[1]: https://github.com/opencontainers/runc/blob/v0.0.4/libcontainer/SPEC.md#runtime-and-init-process

vishh replied:
You are stating runC's behavior. Why do we have to require that behavior in the Spec?

wking replied:
On Mon, Oct 12, 2015 at 01:20:20PM -0700, Vish Kannan wrote:
> You are stating runC's behavior. Why do we have to require that
> behavior in the Spec?

This commit is mostly about describing the runC approach so we have a
non-controversial starting point for future evolution (“is that what
runC does” is something that is harder to disagree on than “should
that be how the spec works” ;).

That being said, I *do* think all runtimes will need to execute
runtime-specified code in the container process [1].

[1]: https://github.com/opencontainers/specs/issues/115#issuecomment-147512709

vishh commented on runtime-config.md:
```diff
@@ -12,27 +12,28 @@ Presently there are `Prestart`, `Poststart` and `Poststop`.
 ### Post-start
 
-The post-start hooks are called after the user process is started.
+The post-start hooks are called [after the user process is started](runtime.md#typical-lifecycle).
```
How about `init process` instead of `user process`?

wking replied:
On Mon, Oct 12, 2015 at 12:48:09PM -0700, Vish Kannan wrote:
> >  
> >  ### Post-start
> >  
> > -The post-start hooks are called after the user process is started.
> > +The post-start hooks are called [after the user process is started](runtime.md#typical-lifecycle).
> 
> How about `init process` instead of `user process`?

The container process (initially running runtime-supplied code) is
also the init process.  Here's how I see the initial communication:

  Host process                Container process
  ----------------------      ------------------------------------
  start
  launches child          ->
  blocks                      joins namespaces
                              unshares namespaces
                              configures unshared namespaces
                              drops permissions, changes user, ...
                          <-  signals that container is complete
  runs pre-start hooks        blocks
  signals hooks complete  ->
                              closes pipe
                              executes user process

So there's a lot going on in the container process before it turns
itself into the “user process”.

wking replied:
On Mon, Oct 12, 2015 at 12:48:09PM -0700, Vish Kannan wrote:
> How about `init process` instead of `user process`?

Also “init process” implies a PID namespace, and those are optional
with this spec.

wking replied:
On Mon, Oct 12, 2015 at 12:56:13PM -0700, W. Trevor King wrote:
>   Host process                Container process
>   ----------------------      ------------------------------------
>   start
>   launches child          ->
>   blocks                      joins namespaces

After launching and before blocking, the host process should also be
adding the container process to its cgroup (if you're using any
cgroups).

vishh replied:
I get the runC behavior @wking. 

One alternative to the runC behavior is that of unsharing namespaces (excepting pid), bind mounting the namespaces, and letting the hooks run before starting a new process in a new pid namespace (if requested by the user). 
In this scenario, the `container process` isn't required.

Similarly, I'm wondering if the new semantics you are introducing are valid in other OSes.
I'd like to solicit feedback from the folks who are working on Spec implementations on other other OS'es like Windows and Solaris.

wking replied:
On Mon, Oct 12, 2015 at 03:48:06PM -0700, Vish Kannan wrote:
> One alternative to the runC behavior is that of unsharing namespaces
> (excepting pid), bind mounting the namespaces, and letting the hooks
> run before starting a new process in a new pid namespace (if
> requested by the user).  In this scenario, the `container process`
> isn't required.

In the “user requested a PID namespace” case, that sounds like “the
container isn't fully setup” (since you don't have the requested PID
namespace at pre-start time).

In either case, there is *still* a container process with
runtime-specified code at pre-start-hook time, because it's holding
open the unshared namespaces.

vishh replied:
Not really. If you bind-mount the ns files, you don't need a process to
hold the namespaces.

On Mon, Oct 12, 2015 at 4:02 PM, W. Trevor King <notifications@github.com>
wrote:

> In runtime-config.md
> <https://github.com/opencontainers/specs/pull/207#discussion_r41810455>:
>
> >
> >  ### Post-start
> >
> > -The post-start hooks are called after the user process is started.
> > +The post-start hooks are called [after the user process is started](runtime.md#typical-lifecycle).
>
> On Mon, Oct 12, 2015 at 03:48:06PM -0700, Vish Kannan wrote: One
> alternative to the runC behavior is that of unsharing namespaces (excepting
> pid), bind mounting the namespaces, and letting the hooks run before
> starting a new process in a new pid namespace (if requested by the user).
> In this scenario, the `container process` isn't required.
> In the “user requested a PID namespace” case, that sounds like “the
> container isn't fully setup” (since you don't have the requested PID
> namespace at pre-start time). In either case, there is *still* a container
> process with runtime-specified code at pre-start-hook time, because it's
> holding open the unshared namespaces.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/specs/pull/207/files#r41810455>.
>

wking replied:
On Mon, Oct 12, 2015 at 04:03:30PM -0700, Vish Kannan wrote:
> Not really. If you bind-mount the ns files, you don't need a process
> to hold the namespaces.

Oh, that will make my life a lot easier :).  Is that:

Host process                    Temporary child
--------------------------      -----------------------------------------
Launch a child process      ->
                                Call clone(2) with CLONE_NEW*
                                Bind mount /proc/self/ns/* somewhere else
                                pivot_root, etc.
                            <-  Exit
Use the bind-mounted files

In that case, excepting the PID namespace from the pre-start
requirements seems like a reasonable prices to pay.  And it will make
explicit ‘create’ and ‘start’ commands possible [1].

On the other hand, we'll probably need to revisit the state JSON that
landed in #87 now that the ‘pid’ entry doesn't make sense for *either*
pre-start or post-stop.

[1]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/frUXLljXy8Y
     Message-ID: <20150915034717.GO18018@odin.tremily.us>

wking replied:
On Mon, Oct 12, 2015 at 04:14:38PM -0700, W. Trevor King wrote:
> Mon, Oct 12, 2015 at 04:03:30PM -0700, Vish Kannan:
> > Not really. If you bind-mount the ns files, you don't need a process
> > to hold the namespaces.
> 
> Oh, that will make my life a lot easier :).  Is that…

And it looks like it is :).  Language supporting this in namespaces(7):

  Bind mounting (see mount(2)) one of the files in this directory to
  somewhere else in the filesystem keeps the corresponding namespace
  of the process specified by pid alive even if all processes
  currently in the namespace terminate.

For example:

  tty1$ unshare --user --uts --map-root-user
  tty1$ hostname alice
  tty1$ echo $$
  25477

  tty2# touch /tmp/uts-alice
  tty2# mount --bind /proc/25477/ns/uts /tmp/uts-alice

  tty1$ exit

  tty2# ps aux | grep 25477 | grep -v grep
  … no hits …
  tty2# nsenter --uts=/tmp/uts-alice hostname
  alice

vishh commented:
I find this PR introducing concepts to the Spec that are implementation details of runC. Input from folks working on implementing the Spec in non-Linux OSes will definitely help.

wking commented:
On Mon, Oct 12, 2015 at 03:50:19PM -0700, Vish Kannan wrote:
> I find this PR introducing concepts to the Spec that are
> implementation details of runC. Input from folks working on
> implementing the Spec in non-Linux OSes will definitely help.

The initial comment for this PR points out that it's just about Linux
containers.  And as I said there, I'm happy to roll in notes for other
platforms if folks know what to write.

And the goal on Linux is to document what runC does now [1], since
that's something that is easier to agree on.

So +1 to feedback like “on Windows, container creation will need
language like …” or “in commit abc, runC is actually cleaning up
containers by doing …”.  But I think feedback like “I think we should
use the unified-hierarchy [2]” should be spun off to the list (it's
mentioned as an option in [3]) and not handled here.

[1]: https://github.com/opencontainers/specs/pull/207#discussion_r41798699
[2]: https://github.com/opencontainers/specs/pull/207#issuecomment-145934135
[3]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/n7vxlhn6Q7g

vishh commented:
@wking: I suggest closing this PR and starting an email thread to solicit feedback from other members.
I hope the existing comments gives you an idea as to what is required to come up with a working proposal.

wking commented:
On Mon, Oct 12, 2015 at 04:57:21PM -0700, Vish Kannan wrote:
> @wking: I suggest closing this PR and starting an email thread to
> solicit feedback from other members.  I hope the existing comments
> gives you an idea as to what is required to come up with a working
> proposal.

Email thread exists [1] (also linked in this PR's original post).  I
moved it to a PR after that thread settled down and I got clearance in
a meeting (see action items in [2]).  I'm happy to continue discussion
in the thread if that's easier for everyone else, but I'm still not
sure what changes you're looking for.

[1]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/frUXLljXy8Y
[2]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2015/opencontainers.2015-09-30-17.00.html

crosbymichael commented:
I think there is still a lot of discussion needed on the lifecycle but this is not an accurate base to start from and will end up being more work if we iterate on this PR instead of discussing first and starting fresh. 

We need to make this detailed but having some of the people who work on runtimes comment as wells as abstract where this works cross platform.  We also need to figure out why we are defining the lifecycle and the types of issues it solves by having this defined or else it will be targeted at a something that is not useful for a consumer.

wking commented:
On Mon, Oct 12, 2015 at 05:16:15PM -0700, Michael Crosby wrote:
> I think there is still a lot of discussion needed on the lifecycle
> but this is not an accurate base to start from and will end up being
> more work if we iterate on this PR instead of discussing first and
> starting fresh.

Fair enough.  @mrunalp volunteered to write one up [1], and this
attempt was because he hadn't gotten around it yet [2].  I'd even be
fine just copying over runC's libcontainer/SPEC.md and then iterating
on that in more tightly-focused PRs.

> We also need to figure out why we are defining the lifecycle and the
> types of issues it solves by having this defined or else it will be
> targeted at a something that is not useful for a consumer.

I think *this* is pretty clear.  For example, it's not currently clear
whether we can rely on the ‘pid’ entry being there in the pre-start
and/or post-stop hooks, what portions of the existing container are
available for pre-start hooks to work with (see #115), what gets
cleaned up when a container exits, what “container processes” even
means [3], what “standard operations” means [4], or what a runtime API
should look like [5].  All of these are difficult discussions to have
without specs for what we think a container lifecycle looks like.

[1]: https://groups.google.com/a/opencontainers.org/d/msg/dev/frUXLljXy8Y/8pH-81moGgAJ
[2]: https://groups.google.com/a/opencontainers.org/d/msg/dev/frUXLljXy8Y/aKFRX56qAAAJ
[3]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/n7vxlhn6Q7g
[4]: https://github.com/opencontainers/specs/pull/209#discussion_r41047185
[5]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/BIxya5eSNLo
